integrated_channels/canvas/client.py

```diff
@@ -67,6 +70,16 @@ def course_create_endpoint(canvas_base_url, canvas_account_id):
             canvas_account_id,
         )
 
+    @staticmethod
+    def course_fetch_endpoint(canvas_base_url, canvas_account_id):
+        """
+        Returns endpoint to GET to for course fetch
+        """
+        return '{}/api/v1/accounts/{}/courses'.format(
```

Comment:
It looks like from here: https://canvas.instructure.com/doc/api/courses.html. that this API is an alias for the more straightforward GET /api/v1/courses/:id.   I would recommend you use this alias instead to keep the canvas account id out of the main logic and as part of the authentication setup only unless there is a good reason.

Reply:
ah yes good callout. The only issue is that we don't have access to the canvas course id :( until we make a course fetch. This is something we should maybe think of doing, perhaps storing the canvas ids on our side would be nice as an association record. But we have not really put a place/table to do that as of yet. e.g. we could store the canvas course id + edx course id association in a table if we wanted sth like

```
customer_lms_configuration_id | edx_course_id | external_lms_course_id
```

or sth like it maybe as a future idea

Reply:
Do we need a course ID for that endpoint? I have a feeling that we can get all courses through `GET /api/v1/courses` but I'm not sure as to the scope of the courses returned. As in, if the `/api/v1/accounts/<account ID>/courses` endpoint actually limits to the sub-account or just returns all courses.

Reply:
side note, are you using this method anywhere?

Reply:
actually, assuming it's for https://github.com/edx/edx-enterprise/pull/1279/files#diff-bd39e800f794c671737477556e9d2d2e5855a2a4d5a83b7c764aba54ec49ece7R350 , I don't think we even need to worry about too many results since we're filtering by course title.

Reply:
I hear you @alex-sheehan-edx on the GET /api/v1/courses, but I checked it again and it only returns more specicifc results, which is weird. The doc also says only for this user. Meaning only courses where the current user (as in, which user? tracked by dev token I think), is enrolled or somehow associated So I could not get it to return the deleted or other course as returned by the search endpoint for example

I agree it would be nice to just call a generic, get me all courses, endpoint, I have not found one! still going to confirm

Reply:
also checking this one https://canvas.instructure.com/doc/api/search.html#method.search.all_courses even though it says 'only courses in public index'

Reply:
and yes, it does return courses across subaccounts! so this resolves one of the issues. The only issue is it does_not_return that deleted course :( if we are to accept that we won't see deleted courses (without using that other search method) and just not want to fetch 'deleted' courses, we can use this one .. I tried {{canvas_url}}/api/v1/search/all_courses?per_page=1000 and I get courses from multiple accounts...


Also if we use this endpoint which may return a large no. of courses, we may need to call it once per `client` lifecycle rather than once per call to `create_content_metadata`

Reply:
there may be one possible way to resolve this


we fetch the root account id (or even just have it stored in canvas config) : I can also try to fetch by calling /api/v1/accounts and using the one with parent_account_id=null.

if we have the root account id, then searching against that id, instead of the subaccount id like I do now, seems to find the course under various subaccount ids!

Reply:
instead of fetching the root account ID, could we just ask the user to supply the root account ID within the configuration?

Reply:
So after going over various options, here is a final idea I have to resolve all of this more cleanly:

```
def find_course_in_canvas(edx_course_id):
  course = find_course_in_account(current_subaccount)
  if not course:
     log(..)
      root_account = find_root_account()
      course = find_course_in_account(root_account) # this may or may not work even
 if not course:
    log(...)
  return course
```

with this, we will still mostly find courses under current subaccount (the most common case), but we will hit the root_account case but if we fail to find a course under current subaccount we will scour at a larger scope under the root account

Reply:
@alex-sheehan-edx this config is already clearly too difficult for users as it is, I would much prefer we use the API call to get it if possible.

Reply:
alrightie! added code to also check under root account id